components/script/dom/gpucommandencoder.rs

```diff
@@ -0,0 +1,114 @@
+
+#[dom_struct]
+pub struct GPUCommandEncoder {
+    reflector_: Reflector,
+    #[ignore_malloc_size_of = "defined in webgpu"]
+    channel: WebGPU,
+    label: DomRefCell<Option<DOMString>>,
+    encoder: WebGPUCommandEncoder,
+}
+
+impl GPUCommandEncoder {
+    pub fn new_inherited(channel: WebGPU, encoder: WebGPUCommandEncoder) -> GPUCommandEncoder {
+        GPUCommandEncoder {
+            channel,
+            reflector_: Reflector::new(),
+            label: DomRefCell::new(None),
+            encoder,
+        }
+    }
+
+    pub fn new(
+        global: &GlobalScope,
+        channel: WebGPU,
+        encoder: WebGPUCommandEncoder,
+    ) -> DomRoot<GPUCommandEncoder> {
+        reflect_dom_object(
+            Box::new(GPUCommandEncoder::new_inherited(channel, encoder)),
+            global,
+            GPUCommandEncoderBinding::Wrap,
+        )
+    }
+}
+
+impl GPUCommandEncoderMethods for GPUCommandEncoder {
+    /// https://gpuweb.github.io/gpuweb/#dom-gpuobjectbase-label
+    fn GetLabel(&self) -> Option<DOMString> {
+        self.label.borrow().clone()
+    }
+
+    /// https://gpuweb.github.io/gpuweb/#dom-gpuobjectbase-label
+    fn SetLabel(&self, value: Option<DOMString>) {
+        *self.label.borrow_mut() = value;
+    }
+
+    /// https://gpuweb.github.io/gpuweb/#dom-gpucommandencoder-begincomputepass
+    fn BeginComputePass(
+        &self,
+        _descriptor: &GPUComputePassDescriptor,
+    ) -> DomRoot<GPUComputePassEncoder> {
+        GPUComputePassEncoder::new(
+            &self.global(),
+            self.channel.clone(),
+            RawPass::new_compute(self.encoder.0),
+        )
+    }
+
+    /// https://gpuweb.github.io/gpuweb/#dom-gpucommandencoder-copybuffertobuffer
+    fn CopyBufferToBuffer(
+        &self,
+        source: &GPUBuffer,
+        source_offset: u64,
+        destination: &GPUBuffer,
+        destination_offset: u64,
+        size: u64,
+    ) {
+        self.channel
+            .0
+            .send(WebGPURequest::CopyBufferToBuffer(
+                self.encoder.0,
+                source.id().0,
+                source_offset,
+                destination.id().0,
+                destination_offset,
+                size,
+            ))
+            .expect("Failed to send CopyBufferToBuffer");
+    }
+
+    /// https://gpuweb.github.io/gpuweb/#dom-gpucommandencoder-finish
+    fn Finish(&self, _descriptor: &GPUCommandBufferDescriptor) -> DomRoot<GPUCommandBuffer> {
+        let (sender, receiver) = ipc::channel().unwrap();
+        self.channel
+            .0
+            .send(WebGPURequest::CommandEncoderFinish(
+                sender,
+                self.encoder.0,
+                // TODO(zakorgy): We should use `_descriptor` here after it's not empty
+                // and the underlying wgpu-core struct is serializable
+            ))
+            .expect("Failed to send CopyBufferToBuffer");
+
+        let buffer = receiver.recv().unwrap();
```

Comment:
similarly to object creation, this shouldn't be blocking

Reply:
I've responded here: https://github.com/servo/servo/pull/25700#discussion_r377018724